Objects/dictobject.c

```diff
@@ -2012,7 +2012,7 @@ dictresize(PyInterpreterState *interp, PyDictObject *mp,
             if (unicode) { // combined unicode -> combined unicode
                 PyDictUnicodeEntry *newentries = DK_UNICODE_ENTRIES(newkeys);
                 if (oldkeys->dk_nentries == numentries && mp->ma_keys->dk_kind == DICT_KEYS_UNICODE) {
-                    memcpy(newentries, oldentries, numentries * sizeof(PyDictUnicodeEntry));
+                    memcpy(newentries, (const void *)oldentries, numentries * sizeof(PyDictUnicodeEntry));
```

Comment:
I don't understand this. The only change is the explicit cast to `const void *`?

Reply:
IIUC this would silence the UBSan misaligned warning, since by definition a `void*` can be read from any location.

But IMHO this is dangerous, because the UBSan warning given in the issue states that a `PyDictUnicodeEntry *` is going to be read from 0x5f7d064233d1, with an alignment of 8 (so this seems to stem from a 64 bit build).

Since this results from `_DK_ENTRIES`, this can only result from `dk_log2_index_bytes=0` due to 

> size_t index = (size_t)1 << dk->dk_log2_index_bytes;

If I read the code correctly, this is only set via `new_keys_object` which has a protective
>assert(log2_size >= PyDict_LOG_MINSIZE);

so that further down in the code `dk_log2_index_bytes` should never be smaller than `PyDict_LOG_MINSIZE=3`.

Hence, I assume this is not a debug build.

Luckily, new_keys_object has only four callers, and two of them pass constants (`PyDict_LOG_MINSIZE` and `NEXT_LOG2_SHARED_KEYS_MAX_SIZE=6`), this leaves `dict_new_presized` and `dictresize`, where the latter unfortunately has  many callers :(

Maybe this helps and hopefully I am not complete mistaken ...

Reply:
Py_EMPTY_KEYS has dk_log2_index_bytes=0. But I don't know why it is passed to dictresize.
Can we get stacktrace for the UBSan error?

Reply:
* When we create an empty dict, we use `Py_EMPTY_KEYS` as the keys object
* When we first insert into that empty dict, we resize it (via `dictresize`)
* In that case `oldkeys` is `Py_EMPTY_KEYS` and `oldentries` is a misaligned `PyDictUnicodeEntry *`
* `memcpy` is called with `numentries=0` and an invalid/misaligned pointer for `oldentries`.
* The UBSan message is confusing: there's not really any misaligned read (it's a zero-sized memcpy), but there's still undefined behavior
* It's undefined behavior to call memcpy with an [invalid pointer](https://en.cppreference.com/w/c/string/byte/memcpy), even if `count` is zero.
* It's also undefined behavior to create the misaligned pointer, even if we don't ever dereference. In other words, calling `DK_UNICODE_ENTRIES()` on `Py_EMPTY_KEYS` is UB.

Reply:
> When we first insert into that empty dict, we resize it (via dictresize)

`insert_to_emptydict()` is optimized for inserting into dict using Py_EMPTY_KEYS.
stacktrace can help to find where `insert_to_emptydict()` can be used.

Anyway, dk_log2_index_bytes in Py_EMPTY_DICT can be 3 instead of 0.

Reply:
```
#0  dictresize (interp=<optimized out>, mp=mp@entry=0x7ffff73a5130, log2_newsize=<optimized out>, unicode=<optimized out>) at Objects/dictobject.c:2015
#1  0x00005555556f7862 in dict_dict_merge (mp=0x7ffff73a5130, other=0x7ffff73a4fb0, override=1, interp=<optimized out>) at Objects/dictobject.c:3740
#2  dict_merge (interp=<optimized out>, a={}, b={'default': '==SUPPRESS==', 'help': 'show this help message and exit', 'dest': 'help', 'option_strings': ['-h', '--help']}, ove
rride=<optimized out>) at Objects/dictobject.c:3812
#3  0x0000555555834776 in _PyEval_EvalFrameDefault (tstate=0x555555c4da00 <_PyRuntime+329472>, frame=<optimized out>, throwflag=0) at Python/generated_cases.c.h:3685
```

> Anyway, dk_log2_index_bytes in Py_EMPTY_DICT can be 3 instead of 0.

That makes sense to me.

Reply:
`dict_dict_merge` has the "fast pass" that just clones from the `b` dict. But the pass can be used only when the `b` is clean and compact.
There is a possible optimization when `a` is empty and `b` is not clean, but I don't know it is worth enough.
Just changing dk_log2_index_bytes is OK for this issue.

Reply:
the 3 is `log2(alignof(void*))`, right?

Reply:
I think the more important reason is that `dk_indices` in `Py_EMPTY_KEYS` is 8 bytes so `dk_log2_index_bytes` should be 3.

If `alignof(void *)` were always 4 (i.e., on 32-bit platforms), then `dk_log2_index_bytes` should *still be* 3.
If `alignof(void *)` were 16, then we'd need more extensive changes (we'd probably need to increase the minimum dictionary size and increase the size of `Py_EMPTY_KEYS`).

https://github.com/python/cpython/blob/67b18a18b66b89e253f38895057ef9f6bae92e7b/Objects/dictobject.c#L603-L604

Reply:
IMHO, for the same reason
> #define PyDict_LOG_MINSIZE 3
> #define PyDict_MINSIZE 8

exist and the many

>assert(log2_size >= PyDict_LOG_MINSIZE);

Thus, I'd suggest to use `PyDict_LOG_MINSIZE` instead of `3` in the pull request?

As Sam said, if there'd ever be the need to support a platform with `alignof(void *)=16`, those two defines will have to be adjusted (or at least `#ifdef` 'ed) and then we won't have to touch the `empty_keys_struct` again.

Reply:
> * It's undefined behavior to call memcpy with an [invalid pointer]

Then, IIUC, we're still in UB land, because due to

https://github.com/python/cpython/blob/67b18a18b66b89e253f38895057ef9f6bae92e7b/Include/internal/pycore_dict.h#L224-L227

we'd "read 0 bytes" from `dk_indices=8`, which is past the `empty_keys_struct`. But at least with correct alignment :-)

If that's really the case, then we could
 *  fill up the `empty_keys_struct` and waste those bytes
 * or skip the `memcpy` if `numentries=0`, but that's an extra if in a hot code path, correct?

 Both options seem a bit too intrusive to  just to workaround the UBSan warning.
 Hopefully there is a better solution.

Reply:
No, it's not an invalid pointer. C allows pointers to the end of an object.

Reply:
Oh, that's great!

Reply:
Sorry to be pesky, I knew this from [C++](https://en.cppreference.com/w/cpp/language/pointer#:~:text=A%20pointer%20past%20the%20end%20of%20an%20object)


But for https://en.cppreference.com/w/c/language/pointer I do not see this or anything similar.

Reply:
See the C11 standard draft, 6.5.6 bullet points 7 and 8. There are probably other references as well.

Reply:
I think this is documented in Section 6.5.9.7 from the C11 (emphasis mine):

> Two pointers compare equal if and only if [...], **both are pointers to one past the last element of the same array object**, or one is a pointer to one past the end of one array object and the other is a pointer to the start of a different array object that happens to immediately follow the first array object in the address space.

Reply:
There are also some examples in https://en.cppreference.com/w/c/language/struct involving structs with flexible array members. The examples are probably not authoritative, but I tend to trust the site:

```
struct s { int n; double d[]; }; // s.d is a flexible array member
double *dp;
...
s2 = malloc(sizeof (struct s) + 6);  // same, but UB to access because 2 bytes are
                                     // missing to complete 1 double
dp = &(s2->d[0]);            // OK, can take address just fine
*dp = 42;                    // undefined behavior
```

Note that constructing `dp` is okay even if evaluating `*dp` is UB.

Reply:
Yeah - flexible arrays to the rescue. Luckily, we have it here in `_dictkeysobject`, which is internal, since in C++ they are not (yet) allowed :)